### ISSUE TITLE
Introduce StreamManager for centralized stream lifecycle management

### ISSUE THREAD

**Issue by boersmamarcel** [boersmamarcel/chatty2#126]
Refactor: Introduce StreamManager for centralized stream lifecycle management
Problem
Stream state is currently scattered across 5 files and 7+ fields:

Location	Field(s)	Purpose
ChattyApp (app_controller.rs:35)	active_stream_tasks: HashMap<String, Task<...>>	Owns GPUI task handles per conversation
ChattyApp (app_controller.rs:40)	pending_task_resolved_ids: HashMap<String, Arc<Mutex<Option<String>>>>	Tracks conversations being created while stream starts
ChattyApp (app_controller.rs:43)	active_create_task: Option<Task<...>>	Guards against duplicate conversation creation
Conversation (conversation.rs)	streaming_message: Option<String>	Accumulated text for background streams
ChatInputState (chat_input.rs:42)	is_streaming: bool	Controls Send/Stop button display
ChatInputState (chat_input.rs:36)	on_stop: Option<StopStreamCallback>	Callback back to ChattyApp
DisplayMessage (message_component.rs)	is_streaming: bool	Per-message streaming indicator
Key issues:

No centralized stream lifecycle management — stream creation, cancellation, and cleanup logic is embedded in ChattyApp methods
UI tightly coupled to stream loop — send_message() is a 500+ line monolith that directly calls chat_view.update() for every chunk
Stop button crashes the app — dropping the GPUI Task mid-execution leaves entity update guards in an inconsistent state
Conversation switching is fragile — complex checks with pending_task_resolved_ids and active_stream_tasks to determine if a background stream is active
Solution: StreamManager
Introduce a centralized StreamManager GPUI Entity that:

Owns all stream lifecycle state in a single HashMap<String, StreamState>
Emits events (StreamManagerEvent) for decoupled UI updates
Uses cancellation tokens (Arc<AtomicBool>) for graceful stream shutdown (fixing the stop crash)
Prepares for future inter-agent communication — streams are independently addressable
StreamManager Design
// src/chatty/services/stream_manager.rs

pub struct StreamState {
    pub response_text: String,
    pub status: StreamStatus,           // Active | Completed | Cancelled | Error(String)
    pub token_usage: Option<(u32, u32)>,
    task: Option<Task<anyhow::Result<()>>>,
    pub trace_json: Option<serde_json::Value>,
    cancel_flag: Arc<AtomicBool>,       // graceful cancellation
}

pub struct StreamManager {
    streams: HashMap<String, StreamState>,
    pending_resolved_ids: HashMap<String, Arc<Mutex<Option<String>>>>,
}

impl EventEmitter<StreamManagerEvent> for StreamManager {}

pub struct GlobalStreamManager {
    pub entity: Option<WeakEntity<StreamManager>>,
}
impl Global for GlobalStreamManager {}

Events — one variant per chunk type, each tagged with conversation_id:

pub enum StreamManagerEvent {
    StreamStarted { conversation_id: String },
    TextChunk { conversation_id: String, text: String },
    ToolCallStarted { conversation_id: String, id: String, name: String },
    ToolCallInput { conversation_id: String, id: String, arguments: String },
    ToolCallResult { conversation_id: String, id: String, result: String },
    ToolCallError { conversation_id: String, id: String, error: String },
    ApprovalRequested { conversation_id: String, id: String, command: String, is_sandboxed: bool },
    ApprovalResolved { conversation_id: String, id: String, approved: bool },
    TokenUsage { conversation_id: String, input_tokens: u32, output_tokens: u32 },
    StreamEnded { conversation_id: String, status: StreamStatus },
}

Key methods:

register_stream(conv_id, task, cancel_flag) — start tracking
register_pending_stream(task, resolved_id, cancel_flag) — for pre-conversation messages
promote_pending(conv_id) — promote __pending__ once conversation is created
handle_chunk(conv_id, StreamChunk) — update internal state + emit event
finalize_stream(conv_id) — mark Completed, drop task
stop_stream(conv_id) → Option<String> — set cancel flag, mark Cancelled, return partial text
is_streaming(conv_id) → bool — single source of truth
get_response_text(conv_id) → Option<&str>
How the UI subscribes
Instead of the stream loop directly calling chat_view.update() 10+ times per chunk type, the UI subscribes to StreamManager events:

// In ChattyApp::setup_callbacks()
cx.subscribe(&stream_manager_entity, |app, _sm, event: &StreamManagerEvent, cx| {
    let chat_view = app.chat_view.clone();
    match event {
        StreamManagerEvent::TextChunk { conversation_id, text } => {
            chat_view.update(cx, |view, cx| {
                if view.conversation_id() == Some(conversation_id) {
                    view.append_assistant_text(text, cx);
                }
            });
        }
        StreamManagerEvent::StreamStarted { conversation_id } => { /* set_streaming(true) */ }
        StreamManagerEvent::StreamEnded { conversation_id, status } => {
            app.handle_stream_ended(conversation_id, status, cx);
        }
        // ... tool calls, approvals
    }
}).detach();

Refactored stream loop (in send_message)
while let Some(chunk) = stream.next().await {
    if cancel_flag.load(Ordering::Relaxed) { break; }
    match chunk {
        Ok(chunk) => {
            if let StreamChunk::Text(ref text) = chunk {
                cx.update_global::<ConversationsStore, _>(|store, _| {
                    if let Some(conv) = store.get_conversation_mut(&conv_id) {
                        conv.append_streaming_content(text);
                    }
                }).ok();
            }
            stream_manager.update(cx, |sm, cx| sm.handle_chunk(&conv_id, chunk, cx)).ok();
        }
        Err(e) => {
            stream_manager.update(cx, |sm, cx| {
                sm.handle_chunk(&conv_id, StreamChunk::Error(e.to_string()), cx);
            }).ok();
            break;
        }
    }
}
stream_manager.update(cx, |sm, cx| sm.finalize_stream(&conv_id, cx)).ok();

Stop button crash fix
The crash happens because drop(task) kills the async block while it's inside a chat_view.update() or cx.update_global() call.

Fix: Cancellation token pattern:

stop_stream() sets cancel_flag.store(true, Ordering::Relaxed)
The stream loop checks the flag on each iteration — exits cleanly before the next update() call
Task drop is a backstop, but the graceful exit usually wins the race
The StreamEnded(Cancelled) event triggers clean UI finalization
Implementation Phases
Phase 1: Create StreamManager (additive, zero breaking changes)
Create stream_manager.rs with all types and methods
Add module to services/mod.rs
Initialize in main.rs
Verify: cargo build && cargo test && cargo clippy
Phase 2: Dual-write (mirror state to StreamManager)
send_message() also registers with StreamManager after spawning task
Stream loop also calls handle_chunk() alongside existing chat_view.update() calls
stop_stream() also calls StreamManager.stop_stream()
Debug logs verify StreamManager state matches ChattyApp state
Files: app_controller.rs
Phase 3: Wire UI to event subscription
Add cx.subscribe(&stream_manager, ...) in setup_callbacks()
Remove direct chat_view.update() calls from stream loop — loop only calls handle_chunk()
Verify: Send messages, switch conversations mid-stream, verify UI updates correctly
Phase 4: Remove duplicate state from ChattyApp
Remove active_stream_tasks, pending_task_resolved_ids fields
Remove clear_streaming_state(), is_streaming(), is_conversation_streaming()
Update load_conversation() to query StreamManager
Task handles now stored exclusively in StreamManager
Files: app_controller.rs
Phase 5: Extract helper methods from send_message
Extract finalize_stream_response(conv_id, response_text, trace_json)
Extract generate_title_if_needed(conv_id)
Extract persist_conversation(conv_id) (shared between send and stop paths)
Wire into StreamEnded event handler
send_message() shrinks from 500+ to ~100-150 lines
Phase 6: Fix stop button with cancellation token
Add cancel_flag: Arc<AtomicBool> to StreamState
stop_stream() sets flag before dropping task
Stream loop checks flag each iteration
Verify: Rapid stop-send-stop sequences don't crash
Files Changed
File	Change
src/chatty/services/stream_manager.rs	New — StreamManager entity, events, state
src/chatty/services/mod.rs	Add module declaration
src/main.rs	Initialize StreamManager at startup
src/chatty/controllers/app_controller.rs	Major refactor — remove stream fields, simplify send_message/stop_stream, add event subscription
src/chatty/views/chat_input.rs	Minor — no structural changes
src/chatty/views/chat_view.rs	No changes — methods called from event handler instead of inline
Future: Inter-agent communication (Phase 7, deferred)
Each stream becomes addressable for cross-conversation communication:

Add stream_id: Uuid to StreamState (a conversation could have multiple sequential streams)
Add send_to_stream(stream_id, message) for injecting messages into an active stream's context
Add StreamManagerEvent::MessageReceived { stream_id, from_conversation_id, content }
Benefits
Single source of truth for stream lifecycle
UI decoupled from streams — reacts to events, not inline calls
Conversation switching is trivial — event handler filters by conversation_id
Stop button is safe — cancellation token provides graceful shutdown
send_message() shrinks from 500+ to ~100-150 lines
Concurrent streams work naturally — per-conversation HashMap
Future inter-agent communication — streams are independently addressable
Phased rollout — each phase is independently testable
print is as verbatim

Refactor: Introduce StreamManager for centralized stream lifecycle management
Problem
Stream state is currently scattered across 5 files and 7+ fields:

Location	Field(s)	Purpose
ChattyApp (app_controller.rs:35)	active_stream_tasks: HashMap<String, Task<...>>	Owns GPUI task handles per conversation
ChattyApp (app_controller.rs:40)	pending_task_resolved_ids: HashMap<String, Arc<Mutex<Option<String>>>>	Tracks conversations being created while stream starts
ChattyApp (app_controller.rs:43)	active_create_task: Option<Task<...>>	Guards against duplicate conversation creation
Conversation (conversation.rs)	streaming_message: Option<String>	Accumulated text for background streams
ChatInputState (chat_input.rs:42)	is_streaming: bool	Controls Send/Stop button display
ChatInputState (chat_input.rs:36)	on_stop: Option<StopStreamCallback>	Callback back to ChattyApp
DisplayMessage (message_component.rs)	is_streaming: bool	Per-message streaming indicator
Key issues:

No centralized stream lifecycle management — stream creation, cancellation, and cleanup logic is embedded in ChattyApp methods
UI tightly coupled to stream loop — send_message() is a 500+ line monolith that directly calls chat_view.update() for every chunk
Stop button crashes the app — dropping the GPUI Task mid-execution leaves entity update guards in an inconsistent state
Conversation switching is fragile — complex checks with pending_task_resolved_ids and active_stream_tasks to determine if a background stream is active
Solution: StreamManager
Introduce a centralized StreamManager GPUI Entity that:

Owns all stream lifecycle state in a single HashMap<String, StreamState>
Emits events (StreamManagerEvent) for decoupled UI updates
Uses cancellation tokens (Arc<AtomicBool>) for graceful stream shutdown (fixing the stop crash)
Prepares for future inter-agent communication — streams are independently addressable
StreamManager Design
// src/chatty/services/stream_manager.rs

pub struct StreamState {
    pub response_text: String,
    pub status: StreamStatus,           // Active | Completed | Cancelled | Error(String)
    pub token_usage: Option<(u32, u32)>,
    task: Option<Task<anyhow::Result<()>>>,
    pub trace_json: Option<serde_json::Value>,
    cancel_flag: Arc<AtomicBool>,       // graceful cancellation
}

pub struct StreamManager {
    streams: HashMap<String, StreamState>,
    pending_resolved_ids: HashMap<String, Arc<Mutex<Option<String>>>>,
}

impl EventEmitter<StreamManagerEvent> for StreamManager {}

pub struct GlobalStreamManager {
    pub entity: Option<WeakEntity<StreamManager>>,
}
impl Global for GlobalStreamManager {}

Events — one variant per chunk type, each tagged with conversation_id:

pub enum StreamManagerEvent {
    StreamStarted { conversation_id: String },
    TextChunk { conversation_id: String, text: String },
    ToolCallStarted { conversation_id: String, id: String, name: String },
    ToolCallInput { conversation_id: String, id: String, arguments: String },
    ToolCallResult { conversation_id: String, id: String, result: String },
    ToolCallError { conversation_id: String, id: String, error: String },
    ApprovalRequested { conversation_id: String, id: String, command: String, is_sandboxed: bool },
    ApprovalResolved { conversation_id: String, id: String, approved: bool },
    TokenUsage { conversation_id: String, input_tokens: u32, output_tokens: u32 },
    StreamEnded { conversation_id: String, status: StreamStatus },
}

Key methods:

register_stream(conv_id, task, cancel_flag) — start tracking
register_pending_stream(task, resolved_id, cancel_flag) — for pre-conversation messages
promote_pending(conv_id) — promote __pending__ once conversation is created
handle_chunk(conv_id, StreamChunk) — update internal state + emit event
finalize_stream(conv_id) — mark Completed, drop task
stop_stream(conv_id) → Option<String> — set cancel flag, mark Cancelled, return partial text
is_streaming(conv_id) → bool — single source of truth
get_response_text(conv_id) → Option<&str>
How the UI subscribes
Instead of the stream loop directly calling chat_view.update() 10+ times per chunk type, the UI subscribes to StreamManager events:

// In ChattyApp::setup_callbacks()
cx.subscribe(&stream_manager_entity, |app, _sm, event: &StreamManagerEvent, cx| {
    let chat_view = app.chat_view.clone();
    match event {
        StreamManagerEvent::TextChunk { conversation_id, text } => {
            chat_view.update(cx, |view, cx| {
                if view.conversation_id() == Some(conversation_id) {
                    view.append_assistant_text(text, cx);
                }
            });
        }
        StreamManagerEvent::StreamStarted { conversation_id } => { /* set_streaming(true) */ }
        StreamManagerEvent::StreamEnded { conversation_id, status } => {
            app.handle_stream_ended(conversation_id, status, cx);
        }
        // ... tool calls, approvals
    }
}).detach();

Refactored stream loop (in send_message)
while let Some(chunk) = stream.next().await {
    if cancel_flag.load(Ordering::Relaxed) { break; }
    match chunk {
        Ok(chunk) => {
            if let StreamChunk::Text(ref text) = chunk {
                cx.update_global::<ConversationsStore, _>(|store, _| {
                    if let Some(conv) = store.get_conversation_mut(&conv_id) {
                        conv.append_streaming_content(text);
                    }
                }).ok();
            }
            stream_manager.update(cx, |sm, cx| sm.handle_chunk(&conv_id, chunk, cx)).ok();
        }
        Err(e) => {
            stream_manager.update(cx, |sm, cx| {
                sm.handle_chunk(&conv_id, StreamChunk::Error(e.to_string()), cx);
            }).ok();
            break;
        }
    }
}
stream_manager.update(cx, |sm, cx| sm.finalize_stream(&conv_id, cx)).ok();

Stop button crash fix
The crash happens because drop(task) kills the async block while it's inside a chat_view.update() or cx.update_global() call.

Fix: Cancellation token pattern:

stop_stream() sets cancel_flag.store(true, Ordering::Relaxed)
The stream loop checks the flag on each iteration — exits cleanly before the next update() call
Task drop is a backstop, but the graceful exit usually wins the race
The StreamEnded(Cancelled) event triggers clean UI finalization
Implementation Phases
Phase 1: Create StreamManager (additive, zero breaking changes)
Create stream_manager.rs with all types and methods
Add module to services/mod.rs
Initialize in main.rs
Verify: cargo build && cargo test && cargo clippy
Phase 2: Dual-write (mirror state to StreamManager)
send_message() also registers with StreamManager after spawning task
Stream loop also calls handle_chunk() alongside existing chat_view.update() calls
stop_stream() also calls StreamManager.stop_stream()
Debug logs verify StreamManager state matches ChattyApp state
Files: app_controller.rs
Phase 3: Wire UI to event subscription
Add cx.subscribe(&stream_manager, ...) in setup_callbacks()
Remove direct chat_view.update() calls from stream loop — loop only calls handle_chunk()
Verify: Send messages, switch conversations mid-stream, verify UI updates correctly
Phase 4: Remove duplicate state from ChattyApp
Remove active_stream_tasks, pending_task_resolved_ids fields
Remove clear_streaming_state(), is_streaming(), is_conversation_streaming()
Update load_conversation() to query StreamManager
Task handles now stored exclusively in StreamManager
Files: app_controller.rs
Phase 5: Extract helper methods from send_message
Extract finalize_stream_response(conv_id, response_text, trace_json)
Extract generate_title_if_needed(conv_id)
Extract persist_conversation(conv_id) (shared between send and stop paths)
Wire into StreamEnded event handler
send_message() shrinks from 500+ to ~100-150 lines
Phase 6: Fix stop button with cancellation token
Add cancel_flag: Arc<AtomicBool> to StreamState
stop_stream() sets flag before dropping task
Stream loop checks flag each iteration
Verify: Rapid stop-send-stop sequences don't crash
Files Changed
File	Change
src/chatty/services/stream_manager.rs	New — StreamManager entity, events, state
src/chatty/services/mod.rs	Add module declaration
src/main.rs	Initialize StreamManager at startup
src/chatty/controllers/app_controller.rs	Major refactor — remove stream fields, simplify send_message/stop_stream, add event subscription
src/chatty/views/chat_input.rs	Minor — no structural changes
src/chatty/views/chat_view.rs	No changes — methods called from event handler instead of inline
Future: Inter-agent communication (Phase 7, deferred)
Each stream becomes addressable for cross-conversation communication:

Add stream_id: Uuid to StreamState (a conversation could have multiple sequential streams)
Add send_to_stream(stream_id, message) for injecting messages into an active stream's context
Add StreamManagerEvent::MessageReceived { stream_id, from_conversation_id, content }
Benefits
Single source of truth for stream lifecycle
UI decoupled from streams — reacts to events, not inline calls
Conversation switching is trivial — event handler filters by conversation_id
Stop button is safe — cancellation token provides graceful shutdown
send_message() shrinks from 500+ to ~100-150 lines
Concurrent streams work naturally — per-conversation HashMap
Future inter-agent communication — streams are independently addressable
Phased rollout — each phase is independently testable
